pytradfri/group.py

```diff
@@ -24,85 +32,116 @@
     ROOT_GROUPS,
 )
 from .error import ColorError
-from .resource import ApiResource
+from .resource import ApiResource, ApiResourceResponse, TypeRaw
+
+if TYPE_CHECKING:
+    from .gateway import Gateway
+
+
+class GroupResponse(ApiResourceResponse):
+    """Represent API response for a blind."""
+
+    color_hex: Optional[str] = Field(alias=ATTR_LIGHT_COLOR_HEX)
+    dimmer: int = Field(alias=ATTR_LIGHT_DIMMER)
+    group_members: Optional[dict[str, Any]] = Field(alias=ATTR_GROUP_MEMBERS)
+    id: int = Field(alias=ATTR_ID)
+    mood_id: str = Field(alias=ATTR_MOOD)
+    state: int = Field(alias=ATTR_DEVICE_STATE)
 
 
 class Group(ApiResource):
     """Represent a group."""
 
-    def __init__(self, gateway, raw):
+    _model_class: type[GroupResponse] = GroupResponse
+    raw: GroupResponse
+
+    def __init__(self, gateway: Gateway, raw: TypeRaw):
         """Create object of class."""
         super().__init__(raw)
+        self.raw = GroupResponse(**raw)
         self._gateway = gateway
 
     @property
-    def path(self):
+    def path(self) -> list[str]:
         """Path."""
-        return [ROOT_GROUPS, self.id]
+        return [ROOT_GROUPS, str(self.id)]
 
     @property
-    def state(self):
+    def state(self) -> bool:
         """Boolean representing the light state of the group."""
-        return self.raw.get(ATTR_DEVICE_STATE) == 1
+        return self.raw.state == 1
 
     @property
-    def dimmer(self):
+    def dimmer(self) -> int:
         """Dimmer value of the group."""
-        return self.raw.get(ATTR_LIGHT_DIMMER)
+        return self.raw.dimmer
 
     @property
-    def hex_color(self):
+    def hex_color(self) -> str | None:
         """Return hex color."""
-        return self.raw.get(ATTR_LIGHT_COLOR_HEX)
+        if self.raw.color_hex:
+            return self.raw.color_hex
+
+        return None
 
     @property
-    def member_ids(self):
-        """Members of this group."""
-        info = self.raw.get(ATTR_GROUP_MEMBERS, {})
+    def member_ids(self) -> list | list[int]:
+        """
+        Members of this group.
+
+        The resonse looks like this:
+        {"15002": {"9003": [65536, 65537, 65538, 65539]}}
+        """
+        info = self.raw.group_members
 
-        if not info or ATTR_HS_LINK not in info:
+        if info is None or ATTR_HS_LINK not in info:
             return []
 
-        return info[ATTR_HS_LINK].get(ATTR_ID, [])
+        member_ids: list[int] = info[ATTR_HS_LINK][ATTR_ID]
+
+        if member_ids:
```

Comment:
I don't think we need this check unless this can be None and we need to change our typing of the response.

Reply:
I'm not sure to be honest but it would make sense that it returns an empty list rather than nothing.

Reply:
Yes, but according to our current typing. It could be an empty list from the beginning but then we don't need this check.

Do we know that this can be None?

Reply:
I don't know if it can be None.

Reply:
I wasn't intending to change the property's behaviour, only rewrite it for the pydantic class.

Reply:
We have changed it already since we're now using `dict[key]` instead of `dict.get` and to access the `ATTR_ID` key. So we have changed the meaning of the optionality, from `ATTR_ID` may be missing to `ATTR_ID` could hold a `None` value.

Reply:
I just created an empty group on my gateway and it looks like this:
```
    {
        "color_hex": null,
        "created_at": 1634990177,
        "dimmer": 0,
        "group_members": {
            "15002": {
                "9003": [
                    65555
                ]
            }
        },
        "id": 131083,
        "mood_id": "0",
        "name": "TRADFRI GROUP 9",
        "ota_update_state": null,
        "state": 0
    },
```

Reply:
Just posted here but I pasted the wrong group.

Reply:
I just did this: created a group in the app and queried the gateway. The group wasn't there. Hit back in the app and list all groups again. The new group is gone.

If I do the same and move a device into the grup it's not deleted and looks like this:
```
    {
        "color_hex": null,
        "created_at": 1646765851,
        "dimmer": 254,
        "group_members": {
            "15002": {
                "9003": [
                    65554
                ]
            }
        },
        "id": 131084,
        "mood_id": "0",
        "name": "Teatar",
        "ota_update_state": null,
        "state": 1
    }
```

The group I moved my device from looks like this:
```
    {
        "color_hex": null,
        "created_at": 1634661538,
        "dimmer": 0,
        "group_members": {
            "15002": {
                "9003": []
            }
        },
        "id": 131082,
        "mood_id": "0",
        "name": "Luftgrupp",
        "ota_update_state": null,
        "state": 0
    },
```

Reply:
So, I'm inclined to say that `group_members` will always have the keys `15002` and `9003`. The latter could be an empty list or a list of `int`.

Reply:
Sounds good! 👍